limits/filetransfer/limits.go

```diff
@@ -0,0 +1,395 @@
+	}
+	if p.limits.FileTransfer.MaxFileSize > 0 &&
+		uint64(offset) >= p.limits.FileTransfer.MaxFileSize {
+		return false
+	} else {
+		return true
+	}
+}
+
+func (p *Permit) PreserveModes(path string, mode os.FileMode) error {
+	if !p.limits.Enabled {
+		return nil
+	}
+
+	if (mode & syscall.S_ISUID) != 0 {
+		mode &= os.ModePerm
+		if p.limits.FileTransfer.Umask != "" {
+			umask, err := strconv.ParseUint(p.limits.FileTransfer.Umask, 8, 32)
+			if err != nil {
+				return err
+			}
+
+			mode = os.ModePerm ^ os.FileMode(uint32(os.ModePerm)&uint32(umask))
+		}
+		mode |= os.ModeSetuid
+	} else {
+		if p.limits.FileTransfer.Umask != "" {
+			umask, err := strconv.ParseUint(p.limits.FileTransfer.Umask, 8, 32)
+			if err != nil {
+				return err
+			}
+
+			mode = os.ModePerm ^ os.FileMode(uint32(os.ModePerm)&uint32(umask))
+		}
+		mode &= os.ModePerm
+	}
+
+	if p.limits.FileTransfer.PreserveMode {
+		return os.Chmod(path, mode)
+	} else {
+		return nil
+	}
+}
+
+func (p *Permit) PreserveOwnerGroup(path string, uid int, gid int) error {
+	if !p.limits.Enabled {
+		return nil
+	}
+
+	forcedSet := false
+	if p.limits.FileTransfer.OwnerPut != "" {
+		u, err := user.Lookup(p.limits.FileTransfer.OwnerPut)
+		if err != nil {
+			return err
+		}
+		uid, err = strconv.Atoi(u.Uid)
+		if err != nil {
+			return err
+		}
+		forcedSet = true
+	}
+	if p.limits.FileTransfer.GroupPut != "" {
+		u, err := user.LookupGroup(p.limits.FileTransfer.GroupPut)
+		if err != nil {
+			return err
+		}
+		gid, err = strconv.Atoi(u.Gid)
+		if err != nil {
+			return err
+		}
+		forcedSet = true
+	}
+	if forcedSet {
+		return os.Chown(path, uid, gid)
+	}
+	if p.limits.FileTransfer.PreserveOwner {
+		return os.Chown(path, uid, gid)
+	} else {
+		return nil
+	}
+}
+
+func updatePerHourCounters() {
+	const one60th = 0.016666666666666666
+	if counterUpdateRunning {
+		counterUpdateStarted <- false
+		return
+	}
+
+	counterUpdateRunning = true
+	counterUpdateStarted <- true
+	//exponential decrease for "avg" over 1 minute: \exp(-dt[s]/(1*60[s]))
+	expWeight1m := math.Exp(-float64(countersUpdateSleepTimeS) * one60th) // / 60.0)
+	//analogy to the uptime weighted avg
+	//expWeight5m := math.Exp(-float64(countersUpdateSleepTimeS) * 0.003333333333333333)  // / (5*60.0))
+	//expWeight15m := math.Exp(-float64(countersUpdateSleepTimeS) * 0.001111111111111111) // / (15*60.0))
+	deviceCounters.bytesReceivedLast1W = 0.0
+	deviceCounters.bytesTransferredLast1W = 0.0
+
+	for counterUpdateRunning {
+		bytesRXLastMinute0 := deviceCounters.bytesReceived
+		bytesTXLastMinute0 := deviceCounters.bytesTransferred
+		bytesRXLastMinute := uint64(0)
+		bytesTXLastMinute := uint64(0)
+		for i := 0; i < int(int64(time.Minute)/int64(countersUpdateSleepTimeS)/int64(time.Second)); i++ {
+			time.Sleep(time.Duration(countersUpdateSleepTimeS) * time.Second)
+			bytesRXLastMinute += deviceCounters.bytesReceived - bytesRXLastMinute0
+			bytesTXLastMinute += deviceCounters.bytesTransferred - bytesTXLastMinute0
```

Comment:
I don't think this is correct.
Every transfer time interval is counted multiple times except the very last.